### PR TITLE
fix(connlib): domains can resolve to same IPs on same gateway

### LINKED ISSUE #5259: bug: 0-size collection in proptest
- https://github.com/firezone/firezone/actions/runs/9386132634/job/25845961145 No random seed
- https://github.com/firezone/firezone/actions/runs/9386848781/job/25848390766 seed ad82010f5fe572646ffded6d071392a4e67ebcdd7c2add2cd26d6433ad5a16e4

### PR DESCRIPTION
**Pull request by conectado**
Currently, the same proxy IP can only ever point to one DNS record. Proxy IPs are given out on a per-connection basis. As a result, if two or more domains resolve to the same IP on the same gateway, previous entries to this domain are lost and return an empty record as a result.

To fix this issue, we now store the set of resources that resolves to this proxy IP instead of just a single resource. An invariant we have to maintain here is that all of these resources must point to the same gateway. This should always be true because proxy IPs are assigned sequentially across all connections and thus the same IP can always point back to the same proxy IP on the same gateway.

Fixes: #5259.

### CODE REVIEW

File: rust/connlib/tunnel/src/client.rs [hunk 1]
```diff
@@ -330,7 +330,7 @@ pub struct ClientState {
     buffered_dns_queries: VecDeque<DnsQuery<'static>>,
 
     /// The (internal) IPs we have assigned for a certain DNS resource.
-    dns_resources_internal_ips: HashMap<IpAddr, DnsResource>,
+    dns_resources_internal_ips: HashMap<IpAddr, HashSet<DnsResource>>,
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/client.rs]:
```suggestion
    dns_resources_internal_ips: HashMap<IpAddr, (GatewayId, HashSet<DnsResource>)>,
```

**Review by thomaseizinger — APPROVED** [review]
LGTM modulo the comment we talked about! Great fix!

File: rust/connlib/tunnel/src/client.rs (hunk 1 above)
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/client.rs]: @conectado Working on this now.

**Comment by thomaseizinger** [comment]
@conectado I updated the PR description and made the changes we discussed. It already has my approval. Feel free to queue for merge whenever.

@jamilbk I think we should release this soon. Current `main` is basically broken for overlapping DNS resources.

**Comment by jamilbk** [comment]
> @conectado I updated the PR description and made the changes we discussed. It already has my approval. Feel free to queue for merge whenever.
> 
> @jamilbk I think we should release this soon. Current `main` is basically broken for overlapping DNS resources.

`main` is also broken for Docker-deployed Gateways. Seems we are invalidating successful candidates causing an ICE timeout after 10s of successful connectivity. See https://firezonehq.slack.com/archives/C04JHRH9CJF/p1717740884116399?thread_ts=1717723682.742359&channel=C04JHRH9CJF&message_ts=1717740884.116399

**Pull request merged**